File: probe/replay.py
"""Replay captured arguments against two versions of a repo and report, per
function, whether behavior is preserved.

Package-aware (each version is a real `git worktree`, so relative imports work)
and input-real (arguments come from probe.capture, not generation). Each version
runs in its own subprocess; observations are compared structurally with the same
soundness rules as the rest of the engine (uncontrolled I/O / threads / opaque
returns -> refuse, don't certify).

Run:  python3 -m probe.replay <repo> <base_ref> <head_ref> <capture.pkl>
"""

import base64
import concurrent.futures
import json
import os
import pickle
import shutil
import subprocess
import sys
import tempfile
from typing import Dict, List, Optional

from . import _procs

# Replay every captured arg-set by default: capping the count silently drops the
# inputs that trigger a divergence (a missed catch), and it doesn't rescue heavy
# functions anyway (they hit the worker timeout regardless). Speed comes from
# parallelism + the per-worker timeout. Users can opt into a cap for speed.
_REPLAY_MAX_ARGS = int(os.environ.get("PROBE_REPLAY_MAX_ARGS", "100000"))
_WORKER_TIMEOUT = int(os.environ.get("PROBE_WORKER_TIMEOUT", "45"))


def _repo_root() -> str:
    return os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


def _git(repo, *args):
    return subprocess.check_output(["git", "-C", repo] + list(args), text=True)


def _worker(worktree, module_name, qualname, blobs, python_exe=None) -> Dict:
    env = dict(os.environ)
    env["PYTHONHASHSEED"] = "0"
    env["PYTHONPATH"] = _repo_root() + os.pathsep + env.get("PYTHONPATH", "")
    job = json.dumps({
        "worktree": worktree, "module_name": module_name,
        "qualname": qualname,
        "args_b64": [base64.b64encode(b).decode("ascii") for b in blobs],
    })
    try:
        proc = _procs.run([python_exe or sys.executable, "-m", "probe._replay_worker"],
                          input=job, capture_output=True, text=True,
                          timeout=_WORKER_TIMEOUT, env=env, cwd=_repo_root())
    except subprocess.TimeoutExpired:
        return {"loaded": False, "error": "timeout", "obs": []}
    if proc.returncode != 0 or not proc.stdout.strip():
        tail = (proc.stderr.strip().splitlines() or ["nonzero exit"])[-1]
        return {"loaded": False, "error": tail, "obs": []}
    return json.loads(proc.stdout)


def _has_opaque(node) -> bool:
    if isinstance(node, list):
        if node and node[0] == "opaque":
            return True
        return any(_has_opaque(x) for x in node)
    return False


def _unsound(obs_list) -> Optional[str]:
    for o in obs_list:
        if o.get("nondet"):
            return "nondeterministic"
        if o.get("io", 0) > 0:
            return "uncontrolled-io"
        if o.get("threads", 0) > 0:
            return "concurrency"
        if "val" in o and _has_opaque(o["val"]):
            return "opaque-return"
        if o.get("self_after") is not None and _has_opaque(o["self_after"]):
            return "opaque-state"
    return None


def _same(a: Dict, b: Dict) -> bool:
    if ("exc" in a) != ("exc" in b):
        return False
    if "exc" in a:
        if a["exc"] != b["exc"]:
            return False
    elif a.get("val") != b.get("val"):
        return False
    return a.get("self_after") == b.get("self_after")  # method mutation is behavior


def _add_worktree(repo, ref) -> str:
    wt = tempfile.mkdtemp(prefix="probe_wt_")
    _git(repo, "worktree", "add", "--detach", wt, ref)
    return wt


def _rm_worktree(repo, wt) -> None:
    try:
        _git(repo, "worktree", "remove", "--force", wt)
    except Exception:
        shutil.rmtree(wt, ignore_errors=True)


def _split_key(key: str):
    """'module::Class.method' -> ('module', 'Class.method')."""
    if "::" in key:
        mod, qual = key.split("::", 1)
        return mod, qual
    return key, key  # legacy/loose form


def _check_key(base_path, head_path, key, blobs, python_exe):
    module_name, qualname = _split_key(key)
    blobs = blobs[:_REPLAY_MAX_ARGS]
    b = _worker(base_path, module_name, qualname, blobs, python_exe)
    h = _worker(head_path, module_name, qualname, blobs, python_exe)
    verdict, note = _verdict(b, h, blobs)
    return (qualname, len(blobs), verdict, note)


def replay_paths(base_path: str, head_path: str, records: Dict[str, List[bytes]],
                 label: str, python_exe: str = None) -> int:
    """Compare two already-materialized versions (directories on disk).

    Per-function checks run in parallel (each spawns two short-lived worker
    subprocesses), and each function replays at most _REPLAY_MAX_ARGS inputs."""
    print("Replay: %s  (%d functions, real captured inputs)" % (label, len(records)))
    print("=" * 74)
    _procs.install()  # main thread: ensure worker subprocesses are reaped on kill

    keys = sorted(records)
    parallelism = min(8, (os.cpu_count() or 2))
    rows = []
    with concurrent.futures.ThreadPoolExecutor(max_workers=parallelism) as ex:
        futures = [ex.submit(_check_key, base_path, head_path, k, records[k], python_exe)
                   for k in keys]
        for fut in concurrent.futures.as_completed(futures):
            rows.append(fut.result())
    rows.sort(key=lambda r: r[0])

    tally: Dict[str, int] = {}
    for _name, _n, verdict, _note in rows:
        tally[verdict] = tally.get(verdict, 0) + 1

    for name, n, verdict, note in rows:
        print("  %-30s n=%-4d %-13s %s" % (name, n, verdict, note))

    checked = sum(1 for _n, _c, v, _t in rows
                  if v in ("equivalent", "divergent", "unverifiable"))
    verifiable = tally.get("equivalent", 0) + tally.get("divergent", 0)
    print("\n" + "-" * 74)
    print("Functions with captured inputs : %d" % len(rows))
    if checked:
        print("Sound auto-verify              : %d/%d = %.0f%%"
              % (verifiable, checked, 100.0 * verifiable / checked))
    print("  equivalent : %d   divergent : %d   unverifiable : %d   not-comparable : %d"
          % (tally.get("equivalent", 0), tally.get("divergent", 0),
             tally.get("unverifiable", 0),
             tally.get("skipped", 0) + tally.get("error", 0)))
    return 1 if tally.get("divergent", 0) else 0


def replay(repo: str, base: str, head: str, capture_path: str,
           python_exe: str = None) -> int:
    with open(capture_path, "rb") as f:
        cap = pickle.load(f)
    records: Dict[str, List[bytes]] = cap["records"]
    base_wt = _add_worktree(repo, base)
    head_wt = repo if head == "WORKTREE" else _add_worktree(repo, head)
    try:
        return replay_paths(base_wt, head_wt, records, "%s..%s" % (base, head),
                            python_exe)
    finally:
        _rm_worktree(repo, base_wt)
        if head_wt != repo:
            _rm_worktree(repo, head_wt)


def _verdict(b: Dict, h: Dict, blobs):
    if b.get("error") or h.get("error"):
        return "error", (b.get("error") or h.get("error"))
    if not b.get("loaded") or not h.get("loaded"):
        return "skipped", "not present in both versions"
    flag = _unsound(b["obs"]) or _unsound(h["obs"])
    if flag:
        return "unverifiable", flag
    if len(b["obs"]) != len(h["obs"]):
        return "error", "observation count mismatch"
    for i, (ob, oh) in enumerate(zip(b["obs"], h["obs"])):
        if not _same(ob, oh):
            try:
                arg = pickle.loads(base64.b64decode(
                    base64.b64encode(blobs[i]).decode("ascii")))
            except Exception:
                arg = "?"
            return "divergent", "@ input #%d %r" % (i, arg)
    return "equivalent", ""


def main(argv=None) -> int:
    argv = list(sys.argv[1:] if argv is None else argv)
    if len(argv) != 4:
        print(__doc__)
        return 2
    repo, base, head, capture_path = argv
    return replay(repo, base, head, capture_path)


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/node/src/replay.js
'use strict';
/**
 * Replay orchestrator: given captured inputs and two version roots, run each
 * captured input on both versions (each in its own subprocess) and produce a
 * per-unit verdict. Mirrors packages/python/probe/replay.py (the verdict model
 * is SPEC/protocol.md section 8).
 */

const path = require('node:path');
const fs = require('node:fs');
const { spawnSync } = require('node:child_process');
const { same, unsound } = require('./soundness');

const WORKER = path.join(__dirname, 'replayWorker.js');

function splitKey(key) {
  const i = key.indexOf('::');
  return [key.slice(0, i), key.slice(i + 2)];
}

function runWorker(versionRoot, moduleRel, qualname, isMethod, argsB64) {
  const job = JSON.stringify({ versionRoot, moduleRel, qualname, is_method: isMethod, args_b64: argsB64 });
  const r = spawnSync(process.execPath, [WORKER], {
    input: job, encoding: 'utf8', maxBuffer: 64 * 1024 * 1024,
    env: { ...process.env, PYTHONHASHSEED: '0' },
  });
  if (r.status !== 0 && !r.stdout) {
    return { loaded: false, error: `worker exited: ${(r.stderr || '').slice(0, 300)}`, obs: [] };
  }
  try { return JSON.parse(r.stdout); } catch (e) {
    return { loaded: false, error: `bad worker output: ${(r.stdout || r.stderr || '').slice(0, 300)}`, obs: [] };
  }
}

function render(o) {
  if (!o) return '∅';
  if ('exc' in o) return `raise ${o.exc}`;
  try { return JSON.stringify(o.val); } catch (e) { return String(o.val); }
}

function verdictFor(base, head) {
  if (base.error || head.error) return { verdict: 'error', note: base.error || head.error };
  if (base.absent || head.absent) return { verdict: 'skipped', note: 'added/removed' };
  if (!base.loaded || !head.loaded) return { verdict: 'skipped', note: 'not loaded in both' };
  const bu = unsound(base.obs);
  if (bu) return { verdict: 'unverifiable', note: bu };
  const hu = unsound(head.obs);
  if (hu) return { verdict: 'unverifiable', note: hu };
  if (base.obs.length !== head.obs.length) return { verdict: 'error', note: 'observation count mismatch' };
  for (let i = 0; i < base.obs.length; i += 1) {
    if (!same(base.obs[i], head.obs[i])) {
      return { verdict: 'divergent', index: i, base: render(base.obs[i]), head: render(head.obs[i]) };
    }
  }
  return { verdict: 'equivalent' };
}

function runReplay(opts) {
  const { capturesFile, beforeRoot, afterRoot } = opts;
  const data = JSON.parse(fs.readFileSync(capturesFile, 'utf8'));
  const byKey = new Map();
  for (const rec of data.records) {
    if (!byKey.has(rec.key)) byKey.set(rec.key, { is_method: rec.is_method, args: [] });
    byKey.get(rec.key).args.push(rec.args_b64);
  }

  const rows = [];
  for (const [key, info] of byKey) {
    const [moduleRel, qualname] = splitKey(key);
    const base = runWorker(beforeRoot, moduleRel, qualname, info.is_method, info.args);
    const head = runWorker(afterRoot, moduleRel, qualname, info.is_method, info.args);
    rows.push({ key, qualname, inputs: info.args.length, ...verdictFor(base, head) });
  }
  return rows;
}

function summarize(rows) {
  const c = { equivalent: 0, divergent: 0, unverifiable: 0, skipped: 0, error: 0 };
  for (const r of rows) c[r.verdict] = (c[r.verdict] || 0) + 1;
  return c;
}

module.exports = { runReplay, summarize, render };
